src/engraving/dom/anchors.cpp

```diff
@@ -0,0 +1,130 @@
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <climits>
+
+#include "anchors.h"
+#include "factory.h"
+#include "score.h"
+#include "spanner.h"
+#include "system.h"
+
+#include "rendering/dev/measurelayout.h"
+
+using namespace mu::engraving::rendering::dev;
+
+namespace mu::engraving {
+/***************************************
+ * EditTimeTickAnchors
+ * ************************************/
+
+void EditTimeTickAnchors::updateAnchors(const EngravingItem* item, Fraction absTick, track_idx_t track)
+{
+    if (!item->allowTimeAnchor()) {
+        item->score()->resetShowAnchors();
+        return;
+    }
+
+    Score* score = item->score();
+    Measure* measure = score->tick2measure(absTick);
+    if (!measure) {
+        return;
+    }
+
+    Fraction measureTick = measure->tick();
+    Measure* prevMeasure = measureTick == absTick ? measure->prevMeasure() : nullptr;
+
+    staff_idx_t staff = track2staff(track);
+    if (prevMeasure) {
+        updateAnchors(prevMeasure, staff);
+    }
+    updateAnchors(measure, staff);
+}
+
+void EditTimeTickAnchors::updateAnchors(Measure* measure, staff_idx_t staffIdx)
+{
+    Fraction startTick = Fraction(0, 1);
+    Fraction endTick = measure->ticks();
+
+    Fraction timeSig = measure->timesig();
+    Fraction halfDivision = Fraction(1, 2 * timeSig.denominator());
+
+    std::set<Fraction> anchorTicks;
+    for (Fraction tick = startTick; tick <= endTick; tick += halfDivision) {
+        anchorTicks.insert(tick);
+    }
+    for (Segment* seg = measure->first(SegmentType::ChordRest); seg; seg = seg->next(SegmentType::ChordRest)) {
+        anchorTicks.insert(seg->rtick());
+    }
+
+    for (const Fraction& anchorTick : anchorTicks) {
+        createTimeTickAnchor(measure, anchorTick, staffIdx);
+    }
+
+    Score* score = measure->score();
+
+    LayoutContext ctx(score);
+    MeasureLayout::layoutTimeTickAnchors(measure, ctx);
+
+    score->updateShowAnchors(staffIdx, measure->tick(), measure->endTick());
+}
+
+Segment* EditTimeTickAnchors::createTimeTickAnchor(Measure* measure, Fraction relTick, staff_idx_t staffIdx)
+{
+    Segment* segment = measure->getSegmentR(SegmentType::TimeTick, relTick);
+
+    track_idx_t track = staff2track(staffIdx);
+    EngravingItem* element = segment->elementAt(track);
+    TimeTickAnchor* anchor = element ? toTimeTickAnchor(element) : new TimeTickAnchor(segment);
```

Comment:
I don't quite get what's going on here
1. we should not rely on elementAt always returning TimeTickAnchor
2. looks like we will add it twice to the segment if it already exists there...
3. it's better to use Factory instead of new

Reply:
1 -> we can rely on that because SegmentType::TimeTick guarantees that only TimeTickAnchor objects can be added to it (like ChordRest with ChordRestType etc)
2, 3 -> fixed, thanks!